ant-networking/src/relay_manager.rs

```diff
@@ -9,47 +9,105 @@
 use crate::driver::{BadNodes, NodeBehaviour};
 use itertools::Itertools;
 use libp2p::{
-    core::transport::ListenerId, multiaddr::Protocol, Multiaddr, PeerId, StreamProtocol, Swarm,
+    core::transport::ListenerId, multiaddr::Protocol, swarm::ConnectionId, Multiaddr, PeerId,
+    StreamProtocol, Swarm,
 };
 use rand::Rng;
-use std::collections::{BTreeMap, HashMap, HashSet, VecDeque};
+use std::{
+    collections::{btree_map::Entry, BTreeMap, HashMap, HashSet, VecDeque},
+    time::SystemTime,
+};
 
 const MAX_CONCURRENT_RELAY_CONNECTIONS: usize = 4;
 const MAX_POTENTIAL_CANDIDATES: usize = 1000;
 
+/// We could get multiple incoming connections from the same peer through multiple relay servers, and only one of them
+/// would succeed. So we wait and collect all such connections peer 'from' peer, instead of just recording the
+///  success/failure for eachconnection.
+const MAX_DURATION_TO_TRACK_INCOMING_CONNECTIONS_PER_PEER: std::time::Duration =
+    std::time::Duration::from_secs(20);
+
+/// The connections from a single peer through multiple relay servers.
+type ConnectionsFromPeer = Vec<(PeerId, ConnectionId, SystemTime, Option<bool>)>;
+
 pub(crate) fn is_a_relayed_peer(addrs: &HashSet<Multiaddr>) -> bool {
     addrs
         .iter()
         .any(|multiaddr| multiaddr.iter().any(|p| matches!(p, Protocol::P2pCircuit)))
 }
 
-/// Manage the relay servers that we are connected to.
+/// Manage the relay servers that a private node is connected to.
 /// This is the client side of the relay server protocol.
 #[derive(Debug)]
 pub(crate) struct RelayManager {
     self_peer_id: PeerId,
-    candidates: VecDeque<(PeerId, Multiaddr)>,
+    /// The potential relay servers that we can connect to.
+    relay_server_candidates: VecDeque<(PeerId, Multiaddr)>,
+    /// The relay servers that we are waiting for a reservation from.
     waiting_for_reservation: BTreeMap<PeerId, Multiaddr>,
-    connected_relays: BTreeMap<PeerId, Multiaddr>,
-
+    /// The relay servers that we are connected to.
+    connected_relay_servers: BTreeMap<PeerId, Multiaddr>,
     /// Tracker for the relayed listen addresses.
     relayed_listener_id_map: HashMap<ListenerId, PeerId>,
+    /// Health of the relayed connections. If the relayed connection is not healthy, we should try to connect to another
+    /// relay.
+    reservation_health: RelayReservationHealth,
+}
+
+#[derive(Debug, Default)]
+struct RelayReservationHealth {
+    /// We could have multiple incoming connections from the same peer through multiple relay servers. But we could
+    /// just have a single one as well.
+    /// This is a map of the 'from peer' to the multiple relay servers that the incoming connections are coming through.
+    incoming_connections_from_remote_peer: BTreeMap<PeerId, ConnectionsFromPeer>,
+    reservation_score: BTreeMap<PeerId, ReservationStat>,
+}
+
+#[derive(Debug, Default, Clone)]
+struct ReservationStat {
```

Comment:
shall this to be the score of the `recent X number attempts` , so that it can reflect the realtime of the health ?